src/Lecture2.hs

```diff
@@ -70,7 +75,14 @@ return the removed element.
 >>> removeAt 10 [1 .. 5]
 (Nothing,[1,2,3,4,5])
 -}
-removeAt = error "TODO"
+removeAt :: Int -> [a] -> (Maybe a, [a])
+removeAt index = go 0 []
+  where
+    go :: Int -> [a] -> [a] -> (Maybe a, [a])
+    go current before rest
+      | null rest        = (Nothing, before ++ rest)
+      | current == index = (Just (head rest), before ++ tail rest)
+      | otherwise        = go (current + 1) (before ++ [head rest]) (tail rest)
```

Comment:
Adding elements to the end of the list is rather slow 🐌

So it's usually better to avoid constructing lists in such a way.

Also, I see that you use `null` and `head + tail` in this function. You can use pattern-matching instead to avoid calls to unsafe functions and guarantee that the code is correct by construction.